src/accelerate/state.py

```diff
@@ -1242,13 +1242,27 @@ def _set_sync_gradients(self, sync_gradients):
 
     def _add_dataloader(self, dataloader):
         "Private function that adds a dataloader to `self.dataloader_references` and sets `in_dataloader` to `True`. Users should not have to call this."
-        self.active_dataloader = dataloader
-        self.dataloader_references.append(self.active_dataloader)
+        self.dataloader_references += [dataloader]
 
     def _remove_dataloader(self, dataloader):
         "Private function that removes a dataloader from `self.dataloader_references` and sets `in_dataloader` to `False` if there are no more dataloaders. Users should not have to call this."
-        self.dataloader_references.remove(dataloader)
-        self.active_dataloader = self.dataloader_references[-1]
+        self.dataloader_references = [
+            dataloader_ref for dataloader_ref in self.dataloader_references if dataloader_ref != dataloader
+        ]
```

Comment:
Let's add comments here on why the code is implemented as is instead of using `append` and `remove`.

Reply:
Definitely - I agree. These would otherwise very easily be refactored if we're not explicit that this is to trigger the property setter correctly.